explore/exploration.h

```diff
@@ -0,0 +1,180 @@
+  {
+    size_t num_actions_scores = scores_last - scores_begin;
+    size_t num_actions_pdf = pdf_last - pdf_first;
+
+	if (num_actions_scores != num_actions_pdf)
+	{
+	#ifdef EXPLORE_NOEXCEPT
+		// fallback to the minimum
+		scores_last = scores_begin + std::min(num_actions_scores, num_actions_pdf);
+		pdf_last = pdf_first + std::min(num_actions_scores, num_actions_pdf);
+  #else    
+		throw std::invalid_argument("length of scores must be equal to length of pdf");
+  #endif
+	}
+
+    if (num_actions_scores == 0)
+      return;
+
+    float norm = 0.;
+    float max_score = *std::max_element(scores_begin, scores_last);
+
+    InputIt s = scores_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && s != scores_last; ++d, ++s)
+    {
+      float prob = exp(lambda*(*s - max_score));
+      norm += prob;
+
+      *d = prob;
+    }
+
+    // normalize
+    for (OutputIt d = pdf_first; d != pdf_last; ++d)
+      *d /= norm;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category scores_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    generate_softmax(lambda, scores_begin, scores_last, scores_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_bag(InputIt top_actions_begin, InputIt top_actions_last, std::input_iterator_tag top_actions_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (pdf_first == pdf_last)
+      return;
+
+    uint32_t num_models = std::accumulate(top_actions_begin, top_actions_last, 0);
+	if (num_models == 0)
+	{
+	#ifdef EXPLORE_NOEXCEPT
+		*pdf_first = 1;
+		return;
+  #else
+		throw std::out_of_range("must supply at least one top_action from a model");
+  #endif
+	}
+
+    // divide late to improve numeric stability
+    InputIt t_a = top_actions_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && t_a != top_actions_last; ++d, ++t_a)
+      *d = *t_a / (float)num_models;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_bag(InputIt top_actions_begin, InputIt top_actions_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category top_actions_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    generate_bag(top_actions_begin, top_actions_last, top_actions_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  // Note: must be inline to compile on Windows VS2017
+  template<typename It>
+  void enforce_minimum_probability(float min_prob, bool update_zero_elements, It pdf_first, It pdf_last, std::random_access_iterator_tag pdf_tag)
```

Comment:
I believe this is still not being used in VW.  Why not?  I'd like to either have safety() or enforce_minimum_probability() (but not both) in the codebase.